### ISSUE TITLE
heap-buffer-overflow in Exiv2::getLong

### ISSUE THREAD

**Issue by gy741** [Exiv2/exiv2#170]
Hello.

I found a heap-buffer-overflow bug in exiv2.

Please confirm.

Thanks.

OS: Ubuntu 17.10 64bit
 Steps to reproduce:
 1.Download the .POC files.
 2.Compile the source code with ASan.
 PoC download : [PoC](https://github.com/gy741/PoC/raw/master/Exiv2::getLong)

```
=================================================================
==22224==ERROR: AddressSanitizer: heap-buffer-overflow on address 0x6020000000b6 at pc 0x7fe02c674057 bp 0x7ffd5ac738f0 sp 0x7ffd5ac738e8
READ of size 1 at 0x6020000000b6 thread T0
    #0 0x7fe02c674056 in Exiv2::getLong(unsigned char const*, Exiv2::ByteOrder) /home/karas/exiv2/src/types.cpp:286:49
    #1 0x7fe02c70583e in Exiv2::Internal::PngChunk::decodeIHDRChunk(Exiv2::DataBuf const&, int*, int*) /home/karas/exiv2/src/pngchunk.cpp:70:22
    #2 0x7fe02c71e9cf in Exiv2::PngImage::readMetadata() /home/karas/exiv2/src/pngimage.cpp:438:21
    #3 0x54a362 in Action::Print::printSummary() /home/karas/exiv2/src/actions.cpp:288:16
    #4 0x5499a8 in Action::Print::run(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) /home/karas/exiv2/src/actions.cpp:240:44
    #5 0x51b98e in main /home/karas/exiv2/src/exiv2.cpp:166:25
    #6 0x7fe02aa831c0 in __libc_start_main /build/glibc-CxtIbX/glibc-2.26/csu/../csu/libc-start.c:308
    #7 0x427789 in _start (/home/karas/exiv2/build/bin/exiv2+0x427789)

0x6020000000b6 is located 1 bytes to the right of 5-byte region [0x6020000000b0,0x6020000000b5)
allocated by thread T0 here:
    #0 0x516e70 in operator new[](unsigned long) (/home/karas/exiv2/build/bin/exiv2+0x516e70)
    #1 0x7fe02c71e64e in Exiv2::DataBuf::DataBuf(long) /home/karas/exiv2/include/exiv2/types.hpp:206:46
    #2 0x7fe02c71e64e in Exiv2::PngImage::readMetadata() /home/karas/exiv2/src/pngimage.cpp:420
    #3 0x54a362 in Action::Print::printSummary() /home/karas/exiv2/src/actions.cpp:288:16
    #4 0x5499a8 in Action::Print::run(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) /home/karas/exiv2/src/actions.cpp:240:44
    #5 0x7fe02aa831c0 in __libc_start_main /build/glibc-CxtIbX/glibc-2.26/csu/../csu/libc-start.c:308

SUMMARY: AddressSanitizer: heap-buffer-overflow /home/karas/exiv2/src/types.cpp:286:49 in Exiv2::getLong(unsigned char const*, Exiv2::ByteOrder)
Shadow bytes around the buggy address:
  0x0c047fff7fc0: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x0c047fff7fd0: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x0c047fff7fe0: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x0c047fff7ff0: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x0c047fff8000: fa fa fd fa fa fa 00 fa fa fa 00 fa fa fa 00 fa
=>0x0c047fff8010: fa fa 00 fa fa fa[05]fa fa fa fa fa fa fa fa fa
  0x0c047fff8020: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c047fff8030: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c047fff8040: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c047fff8050: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c047fff8060: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
Shadow byte legend (one shadow byte represents 8 application bytes):
  Addressable:           00
  Partially addressable: 01 02 03 04 05 06 07 
  Heap left redzone:       fa
  Freed heap region:       fd
  Stack left redzone:      f1
  Stack mid redzone:       f2
  Stack right redzone:     f3
  Stack after return:      f5
  Stack use after scope:   f8
  Global redzone:          f9
  Global init order:       f6
  Poisoned by karas:        f7
  Container overflow:      fc
  Array cookie:            ac
  Intra object redzone:    bb
  ASan internal:           fe
  Left alloca redzone:     ca
  Right alloca redzone:    cb
==22224==ABORTING
```

**Comment by D4N** [comment]
I'll take a look at this soon-ish.

@gy741 I guess you are continuously fuzzing exiv2, aren't you? If that is the case, do you have a dedicated infrastructure for fuzzing?

**Comment by gy741** [comment]
@D4N 

Hello,

I am trying to make Exiv2 software safe.

Using my money, I am taking advantage of cloud services to continue fuzzing.

Thanks.

**Comment by D4N** [comment]
Thanks for your efforts.

If you wish to increase exiv2's security, the best way would be however to contribute patches. The problem is that large parts of exiv2 were not written with security in mind and are susceptible to integer and buffer overflows, so you'll easily find vulnerable code paths with afl and other fuzzers. And while that is good to know, fixing every issue that pops up currently takes up all my time and leaves me with no time to fix the problem at the root. I don't want to be passive aggressive here, but if you could try to fix the issues or identify the root cause (usually it's a integer or buffer overflow or some function returning a null pointer) and report that, you'd help the project a lot more. However if you cannot do that (for whatever reason), please don't stop with your efforts and report the bugs back but please be patient (we are only volunteers working on this in our free time). If you should find exploitable vulnerabilities, please report them directly to me via email (you'll find my address in the commits) and give us some time to fix it before making the bugreport public.

**Comment by gy741** [comment]
@D4N 

Hello,

Thank you for your dedication.

We are volunteers and have limited time.

I am looking for various open source bugs by myself. (a lot of time searching for bug, verifying, and writing reports.)

So, there is insufficient time to analyze the root cause. (Source code analysis and debugging.)

If I use time to analyze the root cause, other bugs will be delayed.

I do not force a quick patch.

Do not feel pressured.

Thanks.

**Comment by piponazo** [comment]
Hi @gy741 , I just tried to reproduce the issue but I could not do it in the **master** branch. 

- Did you have this issue within the master branch ? or were you using a specific tag ?
- Would you mind to try to reproduce the issue on master ?

**Comment by gy741** [comment]
@piponazo 

Hello,

This problem is reproduced on the master branch.

Did you use ASan?

You must use ASan or valgrind.

Thanks.

```
==939== Invalid read of size 1
==939==    at 0x5060E0A: Exiv2::getLong(unsigned char const*, Exiv2::ByteOrder) (in /home/karas/exiv2/lib/libexiv2.so.26.0.0)
==939==    by 0x51034B1: Exiv2::Internal::PngChunk::decodeIHDRChunk(Exiv2::DataBuf const&, int*, int*) (in /home/karas/exiv2/lib/libexiv2.so.26.0.0)
==939==    by 0x50873F8: Exiv2::PngImage::readMetadata() (in /home/karas/exiv2/lib/libexiv2.so.26.0.0)
==939==    by 0x1383D6: Action::Print::printSummary() (in /home/karas/exiv2/bin/exiv2)
==939==    by 0x137FDF: Action::Print::run(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) (in /home/karas/exiv2/bin/exiv2)
==939==    by 0x129564: main (in /home/karas/exiv2/bin/exiv2)
==939==  Address 0x69d5775 is 0 bytes after a block of size 5 alloc'd
==939==    at 0x4C3089F: operator new[](unsigned long) (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==939==    by 0x1482A9: Exiv2::DataBuf::DataBuf(long) (in /home/karas/exiv2/bin/exiv2)
==939==    by 0x50872A7: Exiv2::PngImage::readMetadata() (in /home/karas/exiv2/lib/libexiv2.so.26.0.0)
==939==    by 0x1383D6: Action::Print::printSummary() (in /home/karas/exiv2/bin/exiv2)
==939==    by 0x137FDF: Action::Print::run(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) (in /home/karas/exiv2/bin/exiv2)
==939==    by 0x129564: main (in /home/karas/exiv2/bin/exiv2)
==939== 
==939== Invalid read of size 1
==939==    at 0x5060E1D: Exiv2::getLong(unsigned char const*, Exiv2::ByteOrder) (in /home/karas/exiv2/lib/libexiv2.so.26.0.0)
==939==    by 0x51034B1: Exiv2::Internal::PngChunk::decodeIHDRChunk(Exiv2::DataBuf const&, int*, int*) (in /home/karas/exiv2/lib/libexiv2.so.26.0.0)
==939==    by 0x50873F8: Exiv2::PngImage::readMetadata() (in /home/karas/exiv2/lib/libexiv2.so.26.0.0)
==939==    by 0x1383D6: Action::Print::printSummary() (in /home/karas/exiv2/bin/exiv2)
==939==    by 0x137FDF: Action::Print::run(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) (in /home/karas/exiv2/bin/exiv2)
==939==    by 0x129564: main (in /home/karas/exiv2/bin/exiv2)
==939==  Address 0x69d5776 is 1 bytes after a block of size 5 alloc'd
==939==    at 0x4C3089F: operator new[](unsigned long) (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==939==    by 0x1482A9: Exiv2::DataBuf::DataBuf(long) (in /home/karas/exiv2/bin/exiv2)
==939==    by 0x50872A7: Exiv2::PngImage::readMetadata() (in /home/karas/exiv2/lib/libexiv2.so.26.0.0)
==939==    by 0x1383D6: Action::Print::printSummary() (in /home/karas/exiv2/bin/exiv2)
==939==    by 0x137FDF: Action::Print::run(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) (in /home/karas/exiv2/bin/exiv2)
==939==    by 0x129564: main (in /home/karas/exiv2/bin/exiv2)
==939== 
==939== Invalid read of size 1
==939==    at 0x5060E30: Exiv2::getLong(unsigned char const*, Exiv2::ByteOrder) (in /home/karas/exiv2/lib/libexiv2.so.26.0.0)
==939==    by 0x51034B1: Exiv2::Internal::PngChunk::decodeIHDRChunk(Exiv2::DataBuf const&, int*, int*) (in /home/karas/exiv2/lib/libexiv2.so.26.0.0)
==939==    by 0x50873F8: Exiv2::PngImage::readMetadata() (in /home/karas/exiv2/lib/libexiv2.so.26.0.0)
==939==    by 0x1383D6: Action::Print::printSummary() (in /home/karas/exiv2/bin/exiv2)
==939==    by 0x137FDF: Action::Print::run(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) (in /home/karas/exiv2/bin/exiv2)
==939==    by 0x129564: main (in /home/karas/exiv2/bin/exiv2)
==939==  Address 0x69d5777 is 2 bytes after a block of size 5 alloc'd
==939==    at 0x4C3089F: operator new[](unsigned long) (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==939==    by 0x1482A9: Exiv2::DataBuf::DataBuf(long) (in /home/karas/exiv2/bin/exiv2)
==939==    by 0x50872A7: Exiv2::PngImage::readMetadata() (in /home/karas/exiv2/lib/libexiv2.so.26.0.0)
==939==    by 0x1383D6: Action::Print::printSummary() (in /home/karas/exiv2/bin/exiv2)
==939==    by 0x137FDF: Action::Print::run(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) (in /home/karas/exiv2/bin/exiv2)
==939==    by 0x129564: main (in /home/karas/exiv2/bin/exiv2)
==939== 
```

**Comment by D4N** [comment]
This has been fixed by #230.
